compiler/src/dotty/tools/dotc/core/GadtConstraint.scala

```diff
@@ -2,34 +2,145 @@ package dotty.tools
+    case param: TypeParamRef => reverseMapping(param) match
+      case sym: Symbol => sym.typeRef
+      case null        => param
+    case tp: TypeAlias       => tp.derivedAlias(externalize(tp.alias, theMap))
+    case tp                  => (if theMap == null then ExternalizeMap() else theMap).mapOver(tp)
+
+  private class ExternalizeMap(using Context) extends TypeMap:
+    def apply(tp: Type): Type = externalize(tp, this)(using mapCtx)
+
+  def tvarOrError(sym: Symbol)(using Context): TypeVar =
+    mapping(sym).ensuring(_ != null, i"not a constrainable symbol: $sym").uncheckedNN
+
+  @tailrec final def stripInternalTypeVar(tp: Type): Type = tp match
+    case tv: TypeVar =>
+      val inst = constraint.instType(tv)
+      if inst.exists then stripInternalTypeVar(inst) else tv
+    case _ => tp
+
+  def internalize(tp: Type)(using Context): Type = tp match
+    case nt: NamedType =>
+      val ntTvar = mapping(nt.symbol)
+      if ntTvar == null then tp
+      else ntTvar
+    case _ => tp
+
+  private def containsNoInternalTypes(tp: Type, theAcc: TypeAccumulator[Boolean] | Null = null)(using Context): Boolean = tp match {
+    case tpr: TypeParamRef => !reverseMapping.contains(tpr)
+    case tv: TypeVar => !reverseMapping.contains(tv.origin)
+    case tp =>
+      (if (theAcc != null) theAcc else new ContainsNoInternalTypesAccumulator()).foldOver(true, tp)
+  }
 
-  import dotty.tools.dotc.config.Printers.{gadts, gadtsConstr}
+  private class ContainsNoInternalTypesAccumulator(using Context) extends TypeAccumulator[Boolean] {
+    override def apply(x: Boolean, tp: Type): Boolean = x && containsNoInternalTypes(tp, this)
+  }
 
-  private[core] def getConstraint: Constraint = constraint
-  private[core] def getMapping: SimpleIdentityMap[Symbol, TypeVar] = mapping
-  private[core] def getReverseMapping: SimpleIdentityMap[TypeParamRef, Symbol] = reverseMapping
-  private[core] def getWasConstrained: Boolean = wasConstrained
+  override def toText(printer: Printer): Texts.Text = printer.toText(this)
+
+  /** Provides more information than toText, by showing the underlying Constraint details. */
+  def debugBoundsDescription(using Context): String = i"$this\n$constraint"
+
+  private def copy(
+    myConstraint: Constraint = myConstraint,
+    mapping: SimpleIdentityMap[Symbol, TypeVar] = mapping,
+    reverseMapping: SimpleIdentityMap[TypeParamRef, Symbol] = reverseMapping,
+    wasConstrained: Boolean = wasConstrained,
+  ): GadtConstraint = GadtConstraint(myConstraint, mapping, reverseMapping, wasConstrained)
+end GadtConstraint
+
+object GadtConstraintHandling:
+  def apply(gadt: GadtConstraint): GadtConstraintHandling = new ProperGadtConstraintHandling(gadt)
+
+sealed trait GadtConstraintHandling(private var myGadt: GadtConstraint) {
+  this: ConstraintHandling =>
+
+  def gadt: GadtConstraint              = myGadt
+  private def gadt_=(g: GadtConstraint) = myGadt = g
```

Comment:
Maybe we should not make `GadtConstraint` a private variable in `GadtConstraintHandling`, but declare two methods for retrieving and updating GADT constraints in this trait, and require the class that extends this trait to provide this two methods, just like what we do in `ConstraintHandling` ([here](https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala#L36-L37)). Then `Context.fresh` would do the job of `GadtConstraint.fresh` as long as we put the `GadtConstraint` in the context.